### ISSUE TITLE
New Firefox HAR is missing page.

### ISSUE THREAD

**Issue by soulgalore** [sitespeedio/browsertime#1926]
Hi @juliandescottes I think I have more case where a page is missing:

Running this script:
https://github.com/sitespeedio/dashboard.sitespeed.io/blob/main/tests/docker/desktop/loginWikipedia.js

like `bin/browsertime.js loginWikipedia.js --config secrets.json -n 1 -b firefox` that should generate four pages in. the HAR, but it generates three:
1. Got to the login page
2. Login the user (going to https://en.wikipedia.org/wiki/Main_Page)
3. Barack page
4. President page

But the HAR miss the https://en.wikipedia.org/wiki/Main_Page (that gets redirected when the user login). Easiest way to recreate this is to setup a configuration file with a Wikipedia user (sorry I don't have another way):

```
{
     "wikipedia": {
       "user": "USER,
       "password":"PASWORD"
      } 
}
```

Best
Peter

**Comment by juliandescottes** [comment]
@soulgalore Thanks I'll take a look. Quick question though: as far as I know the current (devtools-based) har builder for firefox is not able to generate multiple pages. Do you use a workaround for that or am I mistaken?

**Comment by soulgalore** [comment]
Hi @juliandescottes you are right. It works like this: For each page we tests, we get the HAR and then we merge them into one: https://github.com/sitespeedio/browsertime/blob/main/lib/support/har/index.js#L236

So in this case where it fails, we click on login to the login page and then Wikipedia redirects the user to  https://en.wikipedia.org/wiki/Main_Page but doesn't seem to generate a new HAR.

**Comment by juliandescottes** [comment]
I intermittently get the same symptoms. One thing I noticed is that when the page entry is missing, all the requests for that page are missing as well, and in the logs I can see

> https://en.wikipedia.org/w/index.php?title=Special:UserLogin&returnto=Main+Page has been tested before within the same run, it will get an extra query parameter named browsertime_run. Make sure to use alias to keep track of the URLs

In successful runs, I don't see this log. I will investigate more tomorrow, but does this ring any bell? Is there anything in the har helper which might confuse browsertime about which URL it is currently testing?

**Comment by juliandescottes** [comment]
Testing locally, it seems the navigation ends too early and stops the recording before any request had time to complete. It would be interesting to check with the old HAR approach to see if the same thing happens.

**Comment by soulgalore** [comment]
> when using --skip-har for info, might not be HAR related? Maybe a Firefox change?

I checked the code and there's a bug there, we only handle a case when we use alias for a URL if we have a HAR. Let me fix that.

I'll fix so we can run the old HAR extension with a switch so it's easier to compare, I'll do that later today.

**Comment by soulgalore** [comment]
Ooops I didn't mean to close it.

@juliandescottes I was wrong on a couple of things: The old HAR exporter didn't produce any error but the HAR is actually faulty, it generates a broken HAR compared to the specification.

I modified script and added an extra wait, it seems like when we click on login, the page complete check happens before the actual navigation, that's why as you said the events doesn't happen. With that fix the HAR pages are generated correctly but it seems like the click on the login button still misses requests? I compared with the HAR I got from Chrome (generated using https://github.com/sitespeedio/chrome-har) and that one has +130 request and the one with bidi har has 30. Not 100% sure though that the Chrome HAR is 100% correct). One thing that would be cool would be if we could turn on logging in the bidi har lib, so that all requests/responses are logged just for debug?

You can change your script to:
```
// Stop and collect the measurement before the next page we want to measure
await commands.wait.byTime(10_000);
await commands.measure.stop();
```

That makes a ten second wait before we say the test is over (I'll fix that later in the Browsertime code).  When I look in the new HAR it is missing the POST to the login page? There's a redirect change happening where cookies are set on different domains and I think maybe we miss some events there.

In the main branch I added back the old HAR exporter and then we can take the time to fix the new without any hurry. To run with the bidiversion you enable it with `--firefox.bidihar`

**Comment by juliandescottes** [comment]
@soulgalore  Having a toggle for the bidi har sounds great while we work out the remaining issues, thanks :) 

I agree we should add logging to the helper. On top of that, I don't know if there's an easy way to get firefox logs when running browsertime, but we could get BiDi logs with `remote.log.level` set to `Trace`. We would directly see all events emitted by BiDi then. But for now I'll add some logging. Probably event name, request id and request url could be enough to avoid making it too verbose. I'll also review cases where I discard requests or events because they have incomplete data and will add extra logging there.

Regarding missing requests between Chrome and Firefox, I know we are missing two types of requests right now: some image requests hitting a specific image cache, and requests blocked by CSP. It sounds odd if that accounts for 100 requests for that scenario though? Do you have an example HAR generated by chrome for this scenario I could look at?

**Comment by soulgalore** [comment]
@juliandescottes attached HAR from bidi and Chrome, checkout page 2. I think the bidi one should include the POST (when you click on the login button right)?

[hars.zip](https://github.com/sitespeedio/browsertime/files/11145827/hars.zip)

**Comment by juliandescottes** [comment]
@soulgalore I just published 0.0.7 which has a small fix for redirected requests and can be used to add logs via `debugLogs: true` in the options you provide to the adapter:
```javascript
    this.har = new adapters.SeleniumBiDiHarRecorder({
      browsingContextIds: [windowId],
      debugLogs: true,
      driver: runner.getDriver()
    });
```

**Comment by soulgalore** [comment]
Ah cool , let me try with beta.

**Comment by juliandescottes** [comment]
Ok, I tried with --firefox-binary pointing to a recent Nightly, and with this I think we have almost all requests accounted for, but 3 requests don't get properly assigned to a page: 
- POST https://en.wikipedia.org/w/index.php?title=Special:UserLogin&returnto=Main+Page
- GET https://en.wikipedia.org/wiki/Special:CentralLogin/start?token...
- GET https://en.wikipedia.org/wiki/Special:CentralLogin/complete?token...

So basically all the redirects which lead to load the Main Page. That's not too surprising. I don't have a good BiDi event implemented yet to monitor the beginning of a navigation so I create the page entries of the har object mostly by finding the first request which seems to be responsible for the load ... 

In case of redirects however I should be able to find the real request which started the redirect chain.

**Comment by soulgalore** [comment]
Hehe, I also tried the same:

<img width="881" alt="Screenshot 2023-04-07 at 22 12 41" src="https://user-images.githubusercontent.com/540757/230673080-22910e3c-be42-4e8c-bef1-5c22bac2f1aa.png">

**Comment by juliandescottes** [comment]
Ahah, yes that's exactly what I had. 

Ok I fixed the redirect issue in 0.0.8, and the 3 requests are now correctly listed in the second page (still testing with Nightly). The number of requests also seems reasonably close to the Chrome HAR shared in this discussion.

Let me know if you spot another inconsistency!

**Comment by soulgalore** [comment]
Did a try now and this looks much better, thanks for the fix @juliandescottes !

**Comment by soulgalore** [comment]
Cool, I released a new version earlier today with 112 :)
